File: lib/opcode.ml
module P = Printf

type t =
  | OpConst
  | OpReturn
  | OpTrue
  | OpFalse
  | OpAdd
  | OpSub
  | OpMul
  | OpDiv
  | OpValue of int

let debug = function
  | OpConst -> "OpConst"
  | OpReturn -> "OpReturn"
  | OpTrue -> "OpTrue"
  | OpFalse -> "OpFalse"
  | OpAdd -> "OpAdd"
  | OpSub -> "OpSub"
  | OpMul -> "OpMul"
  | OpDiv -> "OpDiv"
  | OpValue _ -> "OpValue"
;;

let operand = function
  | OpValue v -> P.sprintf "0x%X" v
  | _ -> ""
;;

File: src/opcode.c
#ifndef OPCODE_H_
#define OPCODE_H_

typedef enum {
    OP_PUSH,
    OP_POP,
    OP_ADD,
    OP_SUB,
    OP_MUL,
    OP_DIV,
    OP_MOD,
    OP_PRINT,
} Op;

typedef struct {
    Op op;
    int value;
} Inst;

#endif // OPCODE_H_
